### PR TITLE
[ValueTracking] Add support for `llvm.vector.reduce.{xor,or,and}` ops.

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for `computeKnownBits` of `llvm.vector.reduce.{or,and}`; NFC**
- **[ValueTracking] Implement `computeKnownBits` for `llvm.vector.reduce.{or,and}`**
- **[ValueTracking] Add tests for `computeKnownBits` of `llvm.vector.reduce.xor`; NFC**
- **[ValueTracking] Implement `computeKnownBits` for `llvm.vector.reduce.xor`**
- **[ValueTracking] Add tests for `isKnownNonZero` of `llvm.vector.reduce.or`; NFC**
- **[ValueTracking] Implement `isKnownNonZero` for `llvm.vector.reduce.or`**

### CODE REVIEW

**Comment by goldsteinn** [comment]
The missing intrins are `fadd` and `fmul` which i don't think are so simple to implement.
Also note that `add`/`mul` non-zero logic might be more complex than is warranted (proofs in the commit). Happy to drop.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -1621,14 +1621,52 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
+        break;
+      case Intrinsic::vector_reduce_mul:
+      case Intrinsic::vector_reduce_add:
+        // We compute the common bits for all elements then apply the reduce op
+        // NumEle times. This is mostly useful for known high zeros.
+        if (auto *VecTy =
+                dyn_cast<FixedVectorType>(I->getOperand(0)->getType())) {
+          computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
+          KnownBits SingleKnown = Known;
+          for (unsigned i = 1, e = VecTy->getNumElements(); i < e; ++i) {
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:1658]: You can reduce the computational complexity from `O(n)` to `O(lgn)` with [exponentiation by squaring](https://en.wikipedia.org/wiki/Exponentiation_by_squaring).

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1658]: yeah, I figured most likely N isn't too big here and this isn't a particularly hot path so it wasn't worth to code complexity, but no real strong opinions.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -5004,6 +5077,15 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
...
     }
+      // reduce min/max will choose an element from one of the vector elements,
+      // so we can infer and class information that is common to all elements.
+    case Intrinsic::vector_reduce_fmax:
+    case Intrinsic::vector_reduce_fmin:
+    case Intrinsic::vector_reduce_fmaximum:
+    case Intrinsic::vector_reduce_fminimum:
+      computeKnownFPClass(II->getArgOperand(0), Known, InterestedClasses,
+                          Depth + 1, Q);
+      break;
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:5088]: Can you please split this into a separate PR for review by FP folks? I'm not sure things are quite as simple as that (e.g., do we need to account for denormal flush like the scalar min/max handling above)?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -2904,6 +2942,41 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
       case Intrinsic::vector_reduce_smax:
       case Intrinsic::vector_reduce_smin:
         return isKnownNonZero(II->getArgOperand(0), Depth, Q);
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2944]: Can't we easily handle "or" reduction for isKnownNonZero()?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 4]
```diff
@@ -2904,6 +2942,41 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
...
+              if (Overflow)
+                break;
+            }
+          }
+
+          if (Overflow)
+            break;
+          return isKnownNonZero(II->getArgOperand(0), Depth, Q);
+        }
+        break;
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2979]: Yeah, I don't think this is worthwhile.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 5]
```diff
@@ -1621,14 +1621,52 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
+              break;
+            if (II->getIntrinsicID() == Intrinsic::vector_reduce_add)
+              Known = KnownBits::computeForAddSub(
+                  /*Add=*/true, /*NSW=*/false, /*NUW=*/false, SingleKnown,
+                  Known);
+            else
+              Known = KnownBits::mul(SingleKnown, Known);
+          }
+        }
+        break;
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:1669]: I don't think this add/mul handling is worthwhile either.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp]: Ill move add/mul stuff to a new PR that most likely wont go anywhere

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp]: Done, see: https://github.com/llvm/llvm-project/pull/88408

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 6]
```diff
@@ -1621,14 +1621,32 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
       case Intrinsic::vector_reduce_smin:
         computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
         break;
+      case Intrinsic::vector_reduce_xor:
+        computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
+        // The zeros common to all vecs are zero in the output.
+        // If the number of elements is odd, then the common ones remain. If the
+        // number of elements is even, then the common ones becomes zeros.
+        if (auto *VecTy =
+                dyn_cast<FixedVectorType>(I->getOperand(0)->getType())) {
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:1640]: Can also handle scalable if you need to know an even count?

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1640]: How? IIUC the only thing we know about scalable is that sizeof(vec) == N * sizeof(scalar) where N >= MinCount. How can we get odd/even?

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:1640]: It's <scale * N * element>, if you know N is even it doesn't matter what scale is

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1640]: Ah, thanks. Done.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 7]
```diff
@@ -1624,14 +1624,29 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
         computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
         break;
+      case Intrinsic::vector_reduce_xor: {
+        computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
+        // The zeros common to all vecs are zero in the output.
+        // If the number of elements is odd, then the common ones remain. If the
+        // number of elements is even, then the common ones becomes zeros.
+        auto *VecTy = cast<VectorType>(I->getOperand(0)->getType());
+        // Even, so the ones become zeros.
+        if (VecTy->getElementCount().isKnownEven()) {
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:1644]: Shouldn't this have an else that sets the ones to unknown? Otherwise what about something like `vscale x 3` which is not known even but might be.

(Actually, I'd just condition the whole xor code on isKnownEven, we don't care about odd-size vectors anyway.)

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1644]: Yeah, you're right. Pushed fix/test.

Its very minimal code complexity to handle odd, so left it in for now (although think I agree its moreso for completeness than practical applicability). LMK if you feel strongly about dropping.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 8]
```diff
@@ -1624,14 +1624,31 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
       case Intrinsic::vector_reduce_smin:
         computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
         break;
+      case Intrinsic::vector_reduce_xor: {
+        computeKnownBits(I->getOperand(0), Known, Depth + 1, Q);
+        // The zeros common to all vecs are zero in the output.
+        // If the number of elements is odd, then the common ones remain. If the
+        // number of elements is even, then the common ones becomes zeros.
+        auto *VecTy = cast<VectorType>(I->getOperand(0)->getType());
+        // Even, so the ones become zeros.
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:1643]: Redundant with the preceding comment?

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 8 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1643]: Guess the idea is the top comment explains what the block will do. The lower comments explain what the next few lines will do. With this minimal code though yeah it does become a bit redundant, but comments are like sex so...
